pdns/recursordist/rec-eventtrace.hh

```diff
@@ -0,0 +1,248 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#pragma once
+
+#include "namespaces.hh"
+#include "misc.hh"
+#include "noinitvector.hh"
+
+#include <protozero/pbf_writer.hpp>
+#include <optional>
+#include <time.h>
+#include <unordered_map>
+#include <variant>
+
+class RecEventTrace
+{
+public:
+  enum EventType : uint8_t
+  {
+    // Don't forget to add a new entry to the table in the .cc file!
+    CustomEvent = 0,
+    RecRecv = 1,
+    PCacheCheck = 2,
+    SyncRes = 3,
+    AnswerSent = 4,
+    LuaGetTag = 100,
+    LuaGetTagFFI = 101,
+    LuaIPFilter = 102,
+    LuaPreRPZ = 103,
+    LuaPreResolve = 104,
+    LuaPreOutQuery = 105,
+    LuaPostResolve = 106,
+    LuaNoData = 107,
+    LuaNXDomain = 108,
+  };
+
+  static const std::unordered_map<EventType, std::string> s_eventNames;
+
+  RecEventTrace()
+  {
+    reset();
+  }
+
+  RecEventTrace(const RecEventTrace& old) :
+    d_events(std::move(old.d_events)),
```

Comment:
I doubt we will be able to move the content of a const object, and I don't believe we should.

Reply:
yes, I messed things up there.

Reply:
Looked at this again. This was an attempt at avoiding the copy the event vector in `handleNewUDPQuestion`. There is a local obect there that is being passed (by value) to `distributeAsyncFunction`. The original object will be reused and so it does not make a lot of sense to keep the vector in the original. 

I will reinstate a regular copy constructor, but it would be nice to avoid the copy of the vector and do a move. But OTOH it is maybe overkill, it will contain very few events at this point in time.